integration-tests/src/test/java/io/opentelemetry/B3PropagationIntegrationTest.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.ContextPropagators;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import io.opentelemetry.extension.trace.propagation.B3Propagator;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.testing.exporter.InMemorySpanExporter;
+import io.opentelemetry.sdk.trace.SdkTracerProvider;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.sdk.trace.export.SimpleSpanProcessor;
+import java.io.IOException;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Stream;
+import org.jetbrains.annotations.Nullable;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.testcontainers.shaded.okhttp3.Call;
+import org.testcontainers.shaded.okhttp3.OkHttpClient;
+import org.testcontainers.shaded.okhttp3.Request;
+import org.testcontainers.shaded.okhttp3.Response;
+import org.testcontainers.shaded.okhttp3.ResponseBody;
+import spark.Service;
+
+/** Integration tests for the B3 propagators, in various configurations. */
+public class B3PropagationIntegrationTest {
+
+  private static final int server1Port = new Random().nextInt(40000) + 1024;
+  private static final int server2Port = new Random().nextInt(40000) + 1024;
+  private static final InMemorySpanExporter spanExporter = InMemorySpanExporter.create();
+
+  private OpenTelemetrySdk clientSdk;
+  private OpenTelemetrySdk server1Sdk;
+  private OpenTelemetrySdk server2Sdk;
+  private Service server1;
+  private Service server2;
+
+  @SuppressWarnings("UnusedMethod")
+  private static Stream<TextMapPropagator> propagators() {
+    return Stream.of(B3Propagator.injectingMultiHeaders(), B3Propagator.injectingSingleHeader());
+  }
+
+  void setup(TextMapPropagator propagator) {
+    setupClient(propagator);
+    setupServer2(propagator);
+    setupServer1(propagator);
+  }
+
+  @AfterEach
+  void shutdown() {
+    if (server2 != null) {
+      server2.stop();
+    }
+    if (server1 != null) {
+      server1.stop();
+    }
+    spanExporter.reset();
+  }
+
+  @ParameterizedTest
+  @MethodSource("propagators")
```

Comment:
BTW personally prefer ArgumentSource for compile safety and to avoid the suppresswarnings

Reply:
:thu